### ISSUE TITLE
rule type_literal_in_constant_pattern in switch cases

### ISSUE THREAD

**Issue by Fernandomr88** [dart-lang/linter#4791]
so I have a function that basically is this

```
get<T>() {
 switch(T) {
  case Person:
   //code
  case Dog:
   //code
  }
}
```

which is called like `get<Person>();`

and by using the notation suggested by this rule (`Person _`), it does not work.

How am I supposed to change my code to be adequate to this linter rule?

_Originally posted by @Fernandomr88 in https://github.com/dart-lang/lints/issues/117#issuecomment-1775789382_

**Comment by lrhn** [comment]
I'd still recommend changing to one of:
- `case const Person:`, or
- `case == Dog:`

**Comment by Fernandomr88** [comment]
> I'd still recommend changing to one of:
> 
> * `case const Person:`, or
> * `case == Dog:`
> 
> Now that type patterns _exist_, it's easy to misread `case Person:` as doing a type check against `Person`.
> 
> And I'd not make exceptions in the lint.

`case const Person:` doesnt work
`case == Dog:` works

but the suggestion made by the linter is either `case const (Person)` or `case Person _`, and both of those are not suitable for this case

**Comment by lrhn** [comment]
This _is_ a false positive for the lint, which suggests that you intended what would be `case Person()` or `case Person _`.

The workaround, to avoid getting that false positive, is `case == Person` or `case const (Person)` - and sadly the parentheses are needed - whichb both compares `Type` objects for equality.

The documentation should probably explain this.

I don't want to disable the lint if we can see that you are switching on `Type` objects. Even in that case, it's better to move to `== Person` as the pattern, because it's more readable.
